src/MLIAP/mliap_descriptor_so3.cpp

```diff
@@ -0,0 +1,298 @@
+      } else if (strcmp(keywd,"nmax") == 0) {
+        nmax = atoi(keyval);
+        nmaxflag = 1;
+      } else if (strcmp(keywd,"lmax") == 0) {
+        lmax = atoi(keyval);
+        lmaxflag = 1;
+      } else if (strcmp(keywd,"alpha") == 0) {
+        alpha = atof(keyval);
+        alphaflag = 1;
+      } else
+        error->all(FLERR,"Incorrect SO3 parameter file");
+    }
+  }
+
+  if (!rcutfacflag || !nelementsflag ||
+    !elementsflag || !radelemflag || !wjelemflag ||
+      !nmaxflag  || !lmaxflag || !alphaflag)
+    error->all(FLERR,"Incorrect SO3 parameter file");
+
+  // construct cutsq
+
+  double cut;
+  cutmax = 0.0;
+  memory->create(cutsq,nelements,nelements,
+    "mliap/descriptor/so3:cutsq");
+  for (int ielem = 0; ielem < nelements; ielem++) {
+
+    cut = 2.0*radelem[ielem]*rcutfac;
+    if (cut > cutmax) cutmax = cut;
+    cutsq[ielem][ielem] = cut*cut;
+
+    for(int jelem = ielem+1; jelem < nelements; jelem++) {
+      cut = (radelem[ielem]+radelem[jelem])*rcutfac;
+      cutsq[ielem][jelem] = cutsq[jelem][ielem] = cut*cut;
+    }
+
+  }
+
+}
+
+void MLIAPDescriptorSO3::compute_descriptors(class MLIAPData* data)
+{
+
+  so3ptr->spectrum(data->nlistatoms,data->numneighs,
+    data->jelems,wjelem,data->rij,nmax, lmax, rcutfac,alpha,
+      data->ndescriptors);
+
+  for (int ii = 0; ii < data->nlistatoms; ii++) {
+    for (int icoeff = 0; icoeff < data->ndescriptors; icoeff++)
+      data->descriptors[ii][icoeff] =
+        so3ptr->m_plist_r[ii*(data->ndescriptors)+icoeff];
+  }
+
+}
+
+void MLIAPDescriptorSO3::compute_forces(class MLIAPData* data)
+{
+  int npairs=0;
+  for (int ii = 0; ii < data->nlistatoms; ii++)
+    npairs+=data->numneighs[ii];
+
+  int totali;
+
+  so3ptr->spectrum_dxdr(data->nlistatoms,data->numneighs,
+    data->jelems,wjelem,data->rij,nmax, lmax, rcutfac,alpha,npairs,
+      data->ndescriptors);
+
+
+  double fij[3];
+  double **f = atom->f;
+
+  int ij = 0;
+
+  for (int ii = 0; ii < data->nlistatoms; ii++) {
+    const int i = data->iatoms[ii];
+    const int ielem = data->ielems[ii];
+
+    // insure rij, inside, wj, and rcutij are of size jnum
+
+    const int jnum = data->numneighs[ii];
+
+    for (int jj = 0; jj < jnum; jj++) {
+      int j = data->jatoms[ij];
+
+      for(int ir=0;ir<3;ir++){
+        fij[ir]=0.0;
+        for(int icoeff=0;icoeff<data->ndescriptors;icoeff++)
+          fij[ir]+=data->betas[ii][icoeff]*so3ptr->m_dplist_r[(ij
+            *(data->ndescriptors)+icoeff)*3+ir];
+      }
+
+      f[i][0] += fij[0];
+      f[i][1] += fij[1];
+      f[i][2] += fij[2];
+      f[j][0] -= fij[0];
+      f[j][1] -= fij[1];
+      f[j][2] -= fij[2];
+
+      // add in global and per-atom virial contributions
+      // this is optional and has no effect on force calculation
+
+      if (data->vflag)
+        data->pairmliap->v_tally(i,j,fij,data->rij[ij]);
+      ij++;
+    }
+
+  }
+}
+
+void MLIAPDescriptorSO3::compute_force_gradients
+  (class MLIAPData* data)
+{
+
```

Comment:
If those functions are intentionally empty, they can be defined in the header. If not, they probably should contain some call to the Error class here.